ehrql/query_engines/base_sql.py

```diff
@@ -719,3 +727,33 @@ def get_sort_conditions(frame):
     # order of priority (i.e. the most recently applied sort gives us the primary sort
     # condition) so we reverse them here
     return tuple(s.sort_by for s in reversed(get_sorts(frame)))
+
+
+def get_cyclic_coalescence(columns):
+    """
+    Given a list of columns, this produces a list of coalescences of all columns with the
+    first input to coalesce at each index being the column at the index in the input columns
+    """
+    # Some SQL engines (sqlite, trino), aggregate functions return NULL if any of the
+    # inputs are NULL. This cyclic coalescence allows us to get rid of any null values, but
+    # retain valid empty strings.
+    # Coalesce returns the first non-null value in a list; by cycling through the list and
+    # calling coalesce with each item as the first value, we end up calling the aggregate
+    # function only on the non-null values
+    # e.g.
+    # cols = ["a", "b", None, ""]
+    # column_cycles = [
+    #   ["a", "b", None, ""],
+    #   ["b", None, "", "a"],
+    #   [None, "", "a", "b"]
+    #   ["", "a", "b", None]
+    # ]
+    # Calling coalesce on each of these returns ["a", "b", "a", "a"] to be used in the
+    # aggregate function
+    len_cols = len(columns)
+    # if there is only one column in the list, just return it. Coalesce requires at least two
+    # arguments
+    if len_cols == 1:
```

Comment:
I don't think so - I think it would error before that, but will check

Reply:
No, this would only happen if you tried to write e.g. `minimum_of()` or `maximum_of()` (with no arguments) in a dataset definition. That does produce a pretty nasty error message, but it happens earlier than this stage, so maybe something to deal with, but not necessarily here.  I made #1508 for it.